### PR TITLE
Fixed potential memory leak during splitting.

### PR DESCRIPTION
**Pull request by alxmrs**
When we open a grib locally, we currently do not use a context manager to open it. This means that `close()` is never called on the open grib data, which would thus leak memory.

In addition to this fix, I've refactored a bit of pipeline code to use less variables.

### CODE REVIEW

**Comment by alxmrs** [comment]
CC: @mahrsee1997 @deepgabani8 Hey! Check out [7ea07dc](https://github.com/google/weather-tools/pull/240/commits/7ea07dcbaec8a6d71d1d9bd4d77f60418194a098) – I think this pattern of closing an XArray dataset in a context manager could be useful to you two (from what we discussed about this morning).

File: weather_sp/splitter_pipeline/file_splitters.py [hunk 1]
```diff
@@ -93,29 +93,31 @@ def split_data(self) -> None:
...
-                            'Variable not found in grib: %s', dim)
-                key = self.output_info.formatted_output_path(splits)
-                if key not in outputs:
-                    outputs[key] = FileSystems.create(key)
-                outputs[key].write(grb.tostring())
-                outputs[key].flush()
-
-            for out in outputs.values():
-                out.close()
+            try:
```
Comment (uhager) [inline weather_sp/splitter_pipeline/file_splitters.py:96]: why is this needed? is  there an indication that is raises an exception here and the output files are left dangling? Or is this just in case? I'd rather not add just-in-case try-blocks if they may not ever be needed.

Comment (alxmrs) [reply weather_sp/splitter_pipeline/file_splitters.py:96]: This offers a guarantee that the `close()` operations are called during any possible error. I don't have evidence that this is a problem now, but it's prudent to include. If I didn't do this, I would refactor this section of code to use a `with` expression (when I tried this, I found that it caused changes of data). 

I wouldn't say that this is never needed: the `finally` block is called every time this method is run -- including when there are exceptions.

File: weather_sp/splitter_pipeline/file_splitters.py [hunk 2]
```diff
@@ -93,29 +93,31 @@ def split_data(self) -> None:
...
+                    out.close()
             self.logger.info('split %s into %d files',
                              self.input_path, len(outputs))
 
     @contextmanager
     def _open_grib_locally(self) -> t.Iterator[t.Iterator[pygrib.gribmessage]]:
         with self._copy_to_local_file() as local_file:
-            yield pygrib.open(local_file.name)
+            with pygrib.open(local_file.name) as gb:
+                yield gb
```
Comment (alxmrs) [inline weather_sp/splitter_pipeline/file_splitters.py:120]: This is the main grib memory leak fix.

File: weather_sp/splitter_pipeline/file_splitters.py (hunk 1 above)
Comment (fredzyda) [reply weather_sp/splitter_pipeline/file_splitters.py:96]: Using with here seems preferable, but I guess this works too.

File: weather_sp/splitter_pipeline/pipeline.py [hunk 3]
```diff
@@ -117,35 +117,31 @@ def run(argv: t.List[str], save_main_session: bool = True):
 
     pipeline_options = PipelineOptions(pipeline_args)
     pipeline_options.view_as(SetupOptions).save_main_session = save_main_session
-    input_pattern = known_args.input_pattern
-    input_base_dir = _get_base_input_directory(input_pattern)
-    output_template = known_args.output_template
-    output_dir = known_args.output_dir
-    formatting = known_args.formatting
+    input_base_dir = _get_base_input_directory(known_args.input_pattern)
```
Comment (fredzyda) [inline weather_sp/splitter_pipeline/pipeline.py:120]: These changes seem somewhat unrelated to the memory leak...

**Review by fredzyda — APPROVED** [review]
Generally looks fine. I had a couple of nits, but nothing to block an important fix on.

File: weather_sp/splitter_pipeline/file_splitters.py (hunk 1 above)
Comment (alxmrs) [reply weather_sp/splitter_pipeline/file_splitters.py:96]: I just stepped into this code through the debugger. I now have a better understanding of what's going on. This pattern now makes sense to me: What I was missing before is that the `write` grib message step happens multiple times per output file. So, we would either have to a) keep a in-memory buffer of all the grib messages that will eventually get flused, then use a `with` expression at the end to write it once, or b) do what we're doing now. 

I'll leave a comment here that tries to explain this, for future reference.

File: weather_sp/splitter_pipeline/pipeline.py (hunk 3 above)
Comment (alxmrs) [reply weather_sp/splitter_pipeline/pipeline.py:120]: Yes, that is right. I'm happy to revert them / move them to a different CL.
